src/Microsoft.Extensions.Options/OptionsCache.cs

```diff
@@ -1,49 +1,65 @@
-// Copyright (c) .NET Foundation. All rights reserved.
+﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
-using System.Threading;
+using System.Collections.Concurrent;
 
 namespace Microsoft.Extensions.Options
 {
-    internal class OptionsCache<TOptions> where TOptions : class, new()
+    /// <summary>
+    /// Used to cache TOptions instances.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public class OptionsCache<TOptions> : IOptionsCache<TOptions> where TOptions : class
     {
-        private readonly Func<TOptions> _createCache;
-        private object _cacheLock = new object();
-        private bool _cacheInitialized;
-        private TOptions _options;
-        private IEnumerable<IConfigureOptions<TOptions>> _setups;
+        private readonly ConcurrentDictionary<string, Lazy<TOptions>> _cache = new ConcurrentDictionary<string, Lazy<TOptions>>(StringComparer.Ordinal);
 
-        public OptionsCache(IEnumerable<IConfigureOptions<TOptions>> setups)
+        public virtual TOptions GetOrAdd(string name, Func<TOptions> createOptions)
         {
-            _setups = setups;
-            _createCache = CreateOptions;
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (createOptions == null)
+            {
+                throw new ArgumentNullException(nameof(createOptions));
+            }
+            return _cache.GetOrAdd(name, new Lazy<TOptions>(createOptions)).Value;
```

Comment:
>new Lazy<TOptions> [](start = 41, length = 18)

Why does this need to be Lazy? GetOrAdd is called with a delegate when you want the TOptions. If the options have not yet been requested, then GetOrAdd will find nothing, and the delegate will be executed to create the options and put them in the dictionary. If the options have already been created, then GetOrAdd will just return them immediately. I'm missing then why Lazy is needed.